ledger/complete/wal/checkpointer.go

```diff
@@ -242,49 +246,129 @@ func CreateCheckpointWriterForFile(dir, filename string) (io.WriteCloser, error)
 	}
 
-	writer := bufio.NewWriter(tmpFile)
+	writer := bufio.NewWriterSize(tmpFile, defaultBufioWriteSize)
 	return &SyncOnCloseRenameFile{
 		file:       tmpFile,
 		targetName: fullname,
 		Writer:     writer,
 	}, nil
 }
 
-// StoreCheckpoint writes the given checkpoint to disk, and also append with a CRC32 file checksum for integrity check.
-func StoreCheckpoint(forestSequencing *flattener.FlattenedForest, writer io.Writer) error {
-	storableNodes := forestSequencing.Nodes
-	storableTries := forestSequencing.Tries
-	header := make([]byte, 4+8+2)
+// StoreCheckpoint writes the given tries to checkpoint file, and also appends
+// a CRC32 file checksum for integrity check.
+// Checkpoint file consists of a flattened forest. Specifically, it consists of:
+//   * a list of encoded nodes, where references to other nodes are by list index.
+//   * a list of encoded tries, each referencing their respective root node by index.
+// Referencing to other nodes by index 0 is a special case, meaning nil.
+//
+// As an important property, the nodes are listed in an order which satisfies
+// Descendents-First-Relationship. The Descendents-First-Relationship has the
+// following important property:
+// When rebuilding the trie from the sequence of nodes, build the trie on the fly,
+// as for each node, the children have been previously encountered.
+// TODO: evaluate alternatives to CRC32 since checkpoint file is many GB in size.
+// TODO: add concurrency if the performance gains are enough to offset complexity.
+func StoreCheckpoint(writer io.Writer, tries ...*trie.MTrie) error {
+
+	var err error
 
 	crc32Writer := NewCRC32Writer(writer)
 
+	// Write header: magic (2 bytes) + version (2 bytes)
+	header := make([]byte, 4)
 	pos := writeUint16(header, 0, MagicBytes)
-	pos = writeUint16(header, pos, VersionV3)
-	pos = writeUint64(header, pos, uint64(len(storableNodes)-1)) // -1 to account for 0 node meaning nil
-	writeUint16(header, pos, uint16(len(storableTries)))
+	_ = writeUint16(header, pos, VersionV4)
 
-	_, err := crc32Writer.Write(header)
+	_, err = crc32Writer.Write(header)
 	if err != nil {
 		return fmt.Errorf("cannot write checkpoint header: %w", err)
 	}
 
-	// 0 element = nil, we don't need to store it
-	for i := 1; i < len(storableNodes); i++ {
-		bytes := flattener.EncodeStorableNode(storableNodes[i])
-		_, err = crc32Writer.Write(bytes)
-		if err != nil {
-			return fmt.Errorf("error while writing node date: %w", err)
+	// assign unique index to every node
+	allNodes := make(map[*node.Node]uint64)
+	allNodes[nil] = 0 // 0th element is nil
+
+	allRootNodes := make([]*node.Node, len(tries))
+
+	// Scratch buffer is used as temporary buffer that node can encode into.
+	// Data in scratch buffer should be copied or used before scratch buffer is used again.
+	// If the scratch buffer isn't large enough, a new buffer will be allocated.
+	// However, 4096 bytes will be large enough to handle almost all payloads
```

Comment:
hmm, not sure if all payloads are expected to fit in 4096 bytes, what if it goes beyond that ?

Reply:
If the 4096 byte scratch buffer isn't large enough, a new buffer will be allocated.  I think 99% of payloads in the mainnet snapshot were smaller than 2048 bytes and I used 4096 as default.